### PR TITLE
ENG-475: trigger reply email

### PR DESCRIPTION
**Pull request by stormcloud266**
- update user model required/optional fields
- regenerate user sdl
- get email submission logic working
- clean up. add email validation
- add relationships in db
- rename id fields. generate sdls for models

### CODE REVIEW

File: api/db/schema.prisma [hunk 1]
```diff
@@ -52,10 +54,22 @@ model Nft {
   contractName String
   data Json
   cloudinaryId String
+  puzzleId String @unique
```
Comment (illepic) [inline api/db/schema.prisma:57]: Multiple `Step`s will have the same `puzzleId`, so this `@unique` annotation will explode.

File: api/db/schema.prisma [hunk 2]
```diff
@@ -52,10 +54,22 @@ model Nft {
...
 
+model Submission {
+  submissionId String @id @default(cuid())
+  createdAt DateTime @default(now())
+  puzzleId String
+  puzzle Puzzle @relation(fields: [puzzleId], references: [puzzleId])
+  user User @relation(fields: [userId], references: [userId])
+  userId String
+  email String?
+  address String?
```
Comment (illepic) [inline api/db/schema.prisma]: Let's change this to be be JSON where the shape can be either `{ email: string; }` or `{address: string; }`

**Review by illepic — CHANGES_REQUESTED** [review]

File: api/db/schema.prisma (hunk 1 above)
Comment (stormcloud266) [reply api/db/schema.prisma:57]: When I remove this `@unique` I get 

`Error parsing attribute "@relation": A one-to-one relation must use unique fields on the defining side. Either add an `@unique` attribute to the field `puzzleId`, or change the relation to one-to-many.`

should I change the puzzle's nft to an array or go about this a different way?
```graphql
model Puzzle {
  ...
  nft Nft[]
}
```

File: api/db/schema.prisma [hunk 3]
```diff
@@ -52,10 +54,22 @@ model Nft {
   contractName String
   data Json
   cloudinaryId String
+  puzzleId String @unique
+  puzzle Puzzle @relation(fields: [puzzleId], references: [puzzleId])
```
Comment (illepic) [inline api/db/schema.prisma:58]:
```
puzzle Puzzle?
```

**Review by illepic — APPROVED** [review]

**Pull request merged**